### ISSUE TITLE
Starting failing supervisor in IEx crashes session

### ISSUE THREAD

**Issue by knightpp** [elixir-lang/elixir#13745]
## Description
IEx Elixir on version 1.17 may fail to catch an error or show it to the user.

## System info
```
$ elixir --version
Erlang/OTP 25 [erts-13.2.2.9] [source] [64-bit] [smp:16:16] [ds:16:16:10] [async-threads:1] [jit:ns]

Elixir 1.17.1 (compiled with Erlang/OTP 25)
```

## How to reproduce

Create a mix project or a script file

`reproduce.exs`
```elixir
defmodule Repr do
  defmodule Work do
    use Task
  end
    
  def run do
    Supervisor.start_link([Work], strategy: :one_for_one)
  end
end

Repr.run()
```

1. Start IEx `iex reproduce.exs`
	- It works and shows the error
```
$ iex reproduce.exs
Erlang/OTP 25 [erts-13.2.2.9] [source] [64-bit] [smp:16:16] [ds:16:16:10] [async-threads:1] [jit:ns]

** (EXIT from #PID<0.105.0>) shutdown: failed to start child: Repr.Work
    ** (EXIT) an exception was raised:
        ** (UndefinedFunctionError) function Repr.Work.start_link/1 is undefined or private
            Repr.Work.start_link([])
            (stdlib 4.3.1.4) supervisor.erl:420: :supervisor.do_start_child_i/3
            (stdlib 4.3.1.4) supervisor.erl:406: :supervisor.do_start_child/2
            (stdlib 4.3.1.4) supervisor.erl:390: anonymous fn/3 in :supervisor.start_children/2
            (stdlib 4.3.1.4) supervisor.erl:1256: :supervisor.children_map/4
            (stdlib 4.3.1.4) supervisor.erl:350: :supervisor.init_children/2
            (stdlib 4.3.1.4) gen_server.erl:851: :gen_server.init_it/2
            (stdlib 4.3.1.4) gen_server.erl:814: :gen_server.init_it/6
```
2. Now run `iex` and run `import_file "reproduce.exs"`
	- expected: see the error
	- actual: the IEx session restarts
```
$ iex
Erlang/OTP 25 [erts-13.2.2.9] [source] [64-bit] [smp:16:16] [ds:16:16:10] [async-threads:1] [jit:ns]

Interactive Elixir (1.17.1) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> import_file "reproduce.exs"

Interactive Elixir (1.17.1) - press Ctrl+C to exit (type h() ENTER for help)  
iex(1)> Repr.run

Interactive Elixir (1.17.1) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> 
```

The same happens when I run `iex -S mix` and executing `Repr.run()`. That does not happen on elixir 1.15/1.16.
```
$ elixir --version
Erlang/OTP 25 [erts-13.2.2.9] [source] [64-bit] [smp:16:16] [ds:16:16:10] [async-threads:1] [jit:ns]

Elixir 1.16.3 (compiled with Erlang/OTP 25)

λ elixir --version
Erlang/OTP 25 [erts-13.2.2.9] [source] [64-bit] [smp:16:16] [ds:16:16:10] [async-threads:1] [jit:ns]

Elixir 1.15.7 (compiled with Erlang/OTP 25)
```

### Operating system

NixOS

### Current behavior

The IEx prints the hello header when crash occurs and does not show any mention of the occurred error.

### Expected behavior

I expect it to produce error explanation.

**Comment by josevalim** [comment]
I can reproduce it like this:

```
Interactive Elixir (1.18.0-dev) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> spawn_link fn -> :omg end
#PID<0.116.0>
iex(2)> spawn_link fn -> raise "oops" end

13:35:59.939 [error] Process #PID<0.117.0> raised an exception
** (RuntimeError) oops
** (EXIT from #PID<0.115.0>) shell process exited with reason: {%RuntimeError{message: "oops"}, []}

Interactive Elixir (1.18.0-dev) - press Ctrl+C to exit (type h() ENTER for help)
iex(2)> spawn_link fn -> exit :shutdown end
```

As you can see, if the link crashes, we let Logger report the error. However, "normal" exits (i.e. `:normal`, `:shutdown`, and `{:shutdown, _}`) are not logged by default. We should choose to either report all IEx crashes or only the "normal" ones. Reporting all is probably better, even though it will lead to duplication in many cases.

**Comment by josevalim** [comment]
Closed by #13753
